### PR TITLE
Code for Best fit, first fit and worst fit algorithm for memory management in OS

### LINKED ISSUE #1876: Add code for Memory Allocation - Best Fit, First Fit and Worst Fit
I would like to add the implementation of this in cpp. Please assign this.
@raghav-dalmia  @MastersAbh  @asha15  @singh-shreya6  @SakshiShreya

### PR DESCRIPTION
**Pull request by Maxz2136**
Fixes #1876 

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
-  Addded Code for First fit memory management algorithm
- Added code for Best fit memory management algorithm
- Added code for Worst fit memory management algorithm

#### Languages Used: 
-  C++

#### Files Added:
- best_fit.cpp
- first_fit.cpp
- worst_fit.cpp

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

File: Memory_management_algorithm/best_fit.cpp [hunk 1]
```diff
@@ -0,0 +1,133 @@
+#include <bits/stdc++.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+using namespace std;
+
+void best_fit(int memory_block[],int process_size[],int m,int n)
+{
+
+	int allocated_list[n][3];
```
Comment (jainaman224) [inline Memory_management_algorithm/best_fit.cpp]: 4 space indentation is not followed.

File: Memory_management_algorithm/best_fit.cpp [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+	int allocation_marking[m];
+
+	for(int i=0;i<m;i++)
+	 allocation_marking[i]=-1;
+
+	for(int i=0;i<n;i++)
```
Comment (jainaman224) [inline Memory_management_algorithm/best_fit.cpp]: Space around operator like `=`, `<` are missing.

File: Memory_management_algorithm/best_fit.cpp [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+		
+	}
+
+	printf("Process Number     Process Size     Allocated memory block size     Memory block number\n");
+
+	for(int i=0;i<n;i++)
+	{
+		if(allocated_list[i][1]!=-1)
+		{
+			printf("%d                         %d                       %d                 %d\n",i,allocated_list[i][0],allocated_list[i][1],allocated_list[i][2]);
```
Comment (jainaman224) [inline Memory_management_algorithm/best_fit.cpp]: Does not look good.

File: Memory_management_algorithm/best_fit.cpp [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+}
+
+
+
+int main()
+{
+
+	int num_of_mem_blocks;
+		int num_of_processes;
```
Comment (jainaman224) [inline Memory_management_algorithm/best_fit.cpp]: Incorrect indentation.

File: Memory_management_algorithm/best_fit.cpp (hunk 1 above)
Comment (Maxz2136) [reply Memory_management_algorithm/best_fit.cpp]: I made the changes @jainaman224

File: Memory_management_algorithm/best_fit.cpp (hunk 3 above)
Comment (Maxz2136) [reply Memory_management_algorithm/best_fit.cpp]: I can't understand what do you mean by "does not look good" ? If you are pointing out the spaces in the **printf**, spaces are provided to format the output properly.

File: Memory_management_algorithm/best_fit.cpp (hunk 4 above)
Comment (Maxz2136) [reply Memory_management_algorithm/best_fit.cpp]: Indentations are properly changed as requested.

File: Memory_management_algorithm/best_fit.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
+#include <bits/stdc++.h>
```
Comment (jainaman224) [inline Memory_management_algorithm/best_fit.cpp]: Create a sub folder best fit.

File: Memory_management_algorithm/best_fit.cpp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+
+using namespace std;
+
+void best_fit(int memory_block[],int process_size[],int m,int n){
+     int allocated_list[n][3];
```
Comment (jainaman224) [inline Memory_management_algorithm/best_fit.cpp]: Use 4 space indentation.

File: Memory_management_algorithm/best_fit.cpp [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+	        if((process_size[i] <= memory_block[j]) && (allocation_marking[j] ==  -1)){
+		    if(memory_block[j] < mem_block){
+		        mem_block = memory_block[j];
+			allocation = j;
+			}
+		}
+	}
+		
+
+       	if(allocation == -1){
```
Comment (jainaman224) [inline Memory_management_algorithm/best_fit.cpp]: Inconsistent indentation.

**Comment by Maxz2136** [comment]
Please review @jainaman224

File: Memory_management_algorithm/best_fit.cpp (hunk 3 above)
Comment (SakshiShreya) [reply Memory_management_algorithm/best_fit.cpp]: Does not look good means these spaces are not looking good. You could have used tabs or something else..

File: Memory_management_algorithm/best_fit.cpp [hunk 8]
```diff
@@ -0,0 +1,102 @@
+#include <bits/stdc++.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+using namespace std;
+
+void best_fit(int memory_block[],int process_size[],int m,int n){
```
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit.cpp]: Add space after ,

File: Memory_management_algorithm/best_fit.cpp [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+	
+	    int allocation = -1;
+	    int mem_block = 10000000;
+
+	    for(int j = 0; j < m; j++){
+	        if((process_size[i] <= memory_block[j]) && (allocation_marking[j] ==  -1)){
+		    if(memory_block[j] < mem_block){
+		        mem_block = memory_block[j];
+			allocation = j;
+			}
```
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit.cpp]: inconsistency in indentation

File: Memory_management_algorithm/best_fit.cpp (hunk 7 above)
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit.cpp]: inconsistency in indentation

File: Memory_management_algorithm/best_fit.cpp [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+       	if(allocation == -1){
+	    allocated_list[i][0] = process_size[i];
+	    allocated_list[i][1] = -1;
+	    allocated_list[i][2] = -1;
+	}
+	else{
+	    allocated_list[i][0] = process_size[i];
+	    allocated_list[i][1] = memory_block[allocation];
+            allocated_list[i][2] = allocation;
```
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit.cpp]: inconsistency in indentation

File: Memory_management_algorithm/best_fit.cpp [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+}
+
+
+int main(){
+
+    int num_of_mem_blocks;
+    int num_of_processes;
+
+    cout<<"Enter the number of memory blocks\n";
```
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit.cpp]: Add space around the operator

File: Memory_management_algorithm/best_fit.cpp [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    cin>>num_of_mem_blocks;
+
+    cout<<"Enter the number of processes\n";
+    cin>>num_of_processes;
+
+    int memory_block[num_of_mem_blocks];
+    int process_size[num_of_processes];
+
+    cout<<"Enter the memory blocks size:\n";
```
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit.cpp]: Add space around the operator

File: Memory_management_algorithm/best_fit.cpp [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+    for(int i = 0; i < num_of_mem_blocks; i++){
+        cin>>memory_block[i];
+    }
+
+    cout<<"Enter the process size:\n";
+    
+    for(int i = 0; i < num_of_processes; i++){
+        cin>>process_size[i];
```
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit.cpp]: Add space around the operator

File: Memory_management_algorithm/best_fit.cpp [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+    int memory_block[num_of_mem_blocks];
+    int process_size[num_of_processes];
+
+    cout<<"Enter the memory blocks size:\n";
+
+    for(int i = 0; i < num_of_mem_blocks; i++){
+        cin>>memory_block[i];
+    }
+
+    cout<<"Enter the process size:\n";
```
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit.cpp]: Add space around the operator

**Comment by somya-kapoor** [comment]
@Maxz2136 Please the required changes in the whole file

File: Memory_management_algorithm/best_fit.cpp (hunk 14 above)
Comment (Maxz2136) [reply Memory_management_algorithm/best_fit.cpp]: Added the spaces around the operator. @somya-kapoor

File: Memory_management_algorithm/best_fit.cpp (hunk 13 above)
Comment (Maxz2136) [reply Memory_management_algorithm/best_fit.cpp]: added the spaces

File: Memory_management_algorithm/best_fit.cpp (hunk 12 above)
Comment (Maxz2136) [reply Memory_management_algorithm/best_fit.cpp]: added the required spaces

File: Memory_management_algorithm/best_fit.cpp (hunk 11 above)
Comment (Maxz2136) [reply Memory_management_algorithm/best_fit.cpp]: added the required spaces

File: Memory_management_algorithm/best_fit.cpp (hunk 8 above)
Comment (Maxz2136) [reply Memory_management_algorithm/best_fit.cpp]: added the spaces after ,

File: Memory_management_algorithm/best_fit.cpp (hunk 3 above)
Comment (Maxz2136) [reply Memory_management_algorithm/best_fit.cpp]: @SakshiShreya Changed the spaces with tabs

File: Memory_management_algorithm/best_fit.cpp [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+	        if((process_size[i] <= memory_block[j]) && (allocation_marking[j] ==  -1)) {
+		    if(memory_block[j] < mem_block) {
+		        mem_block = memory_block[j];
+			allocation = j;
+		    }
+		}
+	    }
+		
+
+       	    if(allocation == -1) {
```
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit.cpp]: inconsistency of indentation

File: Memory_management_algorithm/best_fit.cpp [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+	        allocated_list[i][0] = process_size[i];
+	        allocated_list[i][1] = -1;
+	        allocated_list[i][2] = -1;
+	    }
+	    else {
+	        allocated_list[i][0] = process_size[i];
+	        allocated_list[i][1] = memory_block[allocation];
+                allocated_list[i][2] = allocation;
```
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit.cpp]: inconsistency of indentation

File: Memory_management_algorithm/best_fit.cpp [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+                allocation_marking[allocation] = i;
+	    }
+
+	    printf("Process Number     Process Size     Allocated memory block size     Memory block number\n");
+
+	    for(int i = 0; i < n; i++) {
+	        if(allocated_list[i][1] != -1) {
+	            printf("%d\t\t\t\t%d\t\t%d\t\t\t%d\n",i,allocated_list[i][0],allocated_list[i][1],allocated_list[i][2]);
```
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit.cpp]: Add a space after a comma

File: Memory_management_algorithm/best_fit.cpp [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+	        }
+	        else { 
+	            printf("%d\t\t\t\t%d\t\t\tNot Allocated\t\tNot Allocated\n",i,allocated_list[i][0]);
```
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit.cpp]: Add a space after a comma

File: Memory_management_algorithm/best_fit.cpp [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+        cin >> memory_block[i];
+    }
+
+    cout << "Enter the process size:\n";
+    
+    for(int i = 0; i < num_of_processes; i++) {
+        cin >> process_size[i];
+    }
+
+    best_fit(memory_block,process_size,num_of_mem_blocks,num_of_processes);
```
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit.cpp]: Add a space after a comma

File: Memory_management_algorithm/best_fit/best_fit.cpp [hunk 20]
```diff
@@ -0,0 +1,100 @@
+#include <bits/stdc++.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+using namespace std;
+
+void best_fit(int memory_block[],int process_size[],int m,int n) {
```
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit/best_fit.cpp]: Add a space after a comma

File: Memory_management_algorithm/best_fit/best_fit.cpp [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+        allocation_marking[i] = -1;
+
+	for(int i = 0; i < n; i++) {
+	
+	    int allocation = -1;
+	    int mem_block = 10000000;
+
+	    for(int j = 0; j < m; j++) {
+	        if((process_size[i] <= memory_block[j]) && (allocation_marking[j] ==  -1)) {
+		        if(memory_block[j] < mem_block) {
```
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit/best_fit.cpp]: Use 4 space indentation

File: Memory_management_algorithm/best_fit/best_fit.cpp [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+		            mem_block = memory_block[j];
+			        allocation = j;
+			    }
+		    }
+        }
+
+        if(allocation == -1) {
+            allocated_list[i][0] = process_size[i];
+	        allocated_list[i][1] = -1;
```
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit/best_fit.cpp]: inconsistency of indentation

File: Memory_management_algorithm/best_fit/best_fit.cpp [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+	        allocated_list[i][2] = -1;
+        }
+        else {
+            allocated_list[i][0] = process_size[i];
+	        allocated_list[i][1] = memory_block[allocation];
```
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit/best_fit.cpp]: inconsistency of indentation

File: Memory_management_algorithm/best_fit/best_fit.cpp [hunk 24]
```diff
@@ -0,0 +1,100 @@
...
+            allocated_list[i][2] = allocation;
+            allocation_marking[allocation] = i;
+        }
+
+        printf("Process Number     Process Size     Allocated memory block size     Memory block number\n");
+
+        for(int i = 0; i < n; i++) { 
+            if(allocated_list[i][1] != -1) {
+	            printf("%d                         %d                       %d                 %d\n",i,allocated_list[i][0],allocated_list[i][1],allocated_list[i][2]);
```
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit/best_fit.cpp]: Add a space after comma

File: Memory_management_algorithm/best_fit/best_fit.cpp [hunk 25]
```diff
@@ -0,0 +1,100 @@
...
+	        }
+	        else { 
+	            printf("%d                         %d                  Not Allocated          Not Allocated\n",i,allocated_list[i][0]);
```
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit/best_fit.cpp]: Add a space after comma

**Comment by somya-kapoor** [comment]
@Maxz2136 Please review your code and do similar changes in the whole file.

**Comment by Maxz2136** [comment]
@somya-kapoor I did the changes in the code as requested. Please review the changes.

File: Memory_management_algorithm/first_fit/first_fit.cpp [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+		allocation_marking[i] = -1;
+
+	for (int i = 0; i < n; i++)
+	{
+		for (int j = 0; j < m; j++)
+		{
+			if ((process_size[i] <= memory_block[j]) && (allocation_marking[j] == -1))
+			{
+				allocated_list[i][0] = process_size[i];
+				allocated_list[i][1] = memory_block[j];
```
Comment (somya-kapoor) [inline Memory_management_algorithm/first_fit/first_fit.cpp]: Inconsistency in indentation
use 4 space indentation

File: Memory_management_algorithm/first_fit/first_fit.cpp [hunk 27]
```diff
@@ -0,0 +1,105 @@
...
+		}
+	}
+
+	printf("Process Number     Process Size     Allocated memory block size     Memory block number\n");
+
+	for (int i = 0; i < n; i++)
+	{
+		if (allocated_list[i][1] != -1)
+		{
+			printf("%d\t\t\t%d\t\t\t%d\t\t\t%d\n", i, allocated_list[i][0], allocated_list[i][1], allocated_list[i][2]);
```
Comment (somya-kapoor) [inline Memory_management_algorithm/first_fit/first_fit.cpp]: Inconsistency in indentation
use 4 space indentation

File: Memory_management_algorithm/best_fit.cpp (hunk 15 above)
Comment (somya-kapoor) [reply Memory_management_algorithm/best_fit.cpp]: Inconsistency in indentation
use 4 space indentation

File: Memory_management_algorithm/best_fit.cpp (hunk 16 above)
Comment (somya-kapoor) [reply Memory_management_algorithm/best_fit.cpp]: Inconsistency in indentation
use 4 space indentation

File: Memory_management_algorithm/best_fit.cpp [hunk 28]
```diff
@@ -0,0 +1,102 @@
...
+	    for(int i = 0; i < n; i++) {
+	        if(allocated_list[i][1] != -1) {
+	            printf("%d\t\t\t\t%d\t\t%d\t\t\t%d\n", i, allocated_list[i][0], allocated_list[i][1], allocated_list[i][2]);
+	        }
+	        else { 
+	            printf("%d\t\t\t\t%d\t\t\tNot Allocated\t\tNot Allocated\n", i, allocated_list[i][0]);
+	        }
+		
+	    }
+        }
```
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit.cpp]: Inconsistency in indentation
use 4 space indentation

File: Memory_management_algorithm/first_fit/first_fit.cpp [hunk 29]
```diff
@@ -0,0 +1,105 @@
...
+		else
+		{
+			printf("%d\t\t\t%d\t\t\tNot Allocated\t\tNot Allocated\n", i, allocated_list[i][0]);
+		}
+	}
+}
+
+int main()
+{
+	int num_of_mem_blocks;
```
Comment (somya-kapoor) [inline Memory_management_algorithm/first_fit/first_fit.cpp]: Inconsistency in indentation
use 4 space indentation

File: Memory_management_algorithm/best_fit.cpp (hunk 15 above)
Comment (Maxz2136) [reply Memory_management_algorithm/best_fit.cpp]: The indentation is correct when I am viewing the code in my local editor while viewing it in Github, the indentation is changed. I don't know why it is happening! Does github changes the indentation?

@somya-kapoor

Comment (Maxz2136) [reply Memory_management_algorithm/best_fit.cpp]: Screenshot_2020-03-19 Maxz2136 Algo_Ds_Notes

Comment (Maxz2136) [reply Memory_management_algorithm/best_fit.cpp]: The above screenshot is when I see the code in the editor present in Github itself, the indentation seems correct but when viewed outside of editor, the indentation gets disturbed.

File: Memory_management_algorithm/best_fit.cpp (hunk 5 above)
Comment (somya-kapoor) [inline Memory_management_algorithm/best_fit.cpp]: Change the file name to Memory_Management_Algorithm/Best_Fit.cpp 
Make a similar change in the whole files

File: Memory_management_algorithm/best_fit.cpp (hunk 15 above)
Comment (somya-kapoor) [reply Memory_management_algorithm/best_fit.cpp]: you are using the tab indentation,
use space indentation

**Comment by somya-kapoor** [comment]
@Maxz2136 Use space indentation instead of tab indentation

File: Memory_management_algorithm/best_fit.cpp (hunk 5 above)
Comment (somya-kapoor) [reply Memory_management_algorithm/best_fit.cpp]: Memory_Management_Algorithm/Best_Fit.cpp

File: Memory_management_algorithm/Best_fit.cpp [hunk 30]
```diff
@@ -0,0 +1,102 @@
...
+
+               if(allocation == -1) {
+            allocated_list[i][0] = process_size[i];
+            allocated_list[i][1] = -1;
+            allocated_list[i][2] = -1;
+        }
+        else {
+            allocated_list[i][0] = process_size[i];
+            allocated_list[i][1] = memory_block[allocation];
+                allocated_list[i][2] = allocation;
```
Comment (somya-kapoor) [inline Memory_management_algorithm/Best_fit.cpp]: Inconsistency in indentation

File: Memory_management_algorithm/Best_fit.cpp [hunk 31]
```diff
@@ -0,0 +1,102 @@
...
+            if((process_size[i] <= memory_block[j]) && (allocation_marking[j] ==  -1)) {
+            if(memory_block[j] < mem_block) {
+                mem_block = memory_block[j];
+            allocation = j;
+            }
+        }
+        }
+        
+
+               if(allocation == -1) {
```
Comment (somya-kapoor) [inline Memory_management_algorithm/Best_fit.cpp]: Inconsistency in indentation

File: Memory_management_algorithm/Best_fit.cpp [hunk 32]
```diff
@@ -0,0 +1,102 @@
...
+        int mem_block = 10000000;
+
+        for(int j = 0; j < m; j++) {
+            if((process_size[i] <= memory_block[j]) && (allocation_marking[j] ==  -1)) {
+            if(memory_block[j] < mem_block) {
+                mem_block = memory_block[j];
+            allocation = j;
+            }
+        }
+        }
```
Comment (somya-kapoor) [inline Memory_management_algorithm/Best_fit.cpp]: Inconsistency in indentation

File: Memory_management_algorithm/Best_fit.cpp [hunk 33]
```diff
@@ -0,0 +1,102 @@
...
+        for(int i = 0; i < n; i++) {
+            if(allocated_list[i][1] != -1) {
+                printf("%d\t\t\t\t%d\t\t%d\t\t\t%d\n", i, allocated_list[i][0], allocated_list[i][1], allocated_list[i][2]);
+            }
+            else { 
+                printf("%d\t\t\t\t%d\t\t\tNot Allocated\t\tNot Allocated\n", i, allocated_list[i][0]);
+            }
+        
+        }
+        }
```
Comment (somya-kapoor) [inline Memory_management_algorithm/Best_fit.cpp]: Inconsistency in indentation

File: Memory_management_algorithm/first_fit/first_fit.cpp (hunk 26 above)
Comment (somya-kapoor) [reply Memory_management_algorithm/first_fit/first_fit.cpp]: Inconsistency in indentation
use 4 space indentation

**Comment by somya-kapoor** [comment]
@Maxz2136 Are you working on it?

**Comment by Maxz2136** [comment]
@somya-kapoor I am working on this but I am having correcting the indentation. The indentation seems correct in my editor but when I push it to the repo, in Github, the indentation gets changed.

**Comment by Maxz2136** [comment]
@somya-kapoor I am working on this but I am having trouble maintaining the indentation. The code has correct indentation in my editor but when I push it to the repo, in Github, the indentation gets changed every time.

**Comment by somya-kapoor** [comment]
@Maxz2136 Change the indent type to space and indent size to 4 in Github file

File: Memory_management_algorithm/Best_fit.cpp [hunk 34]
```diff
@@ -0,0 +1,103 @@
+#include <bits/stdc++.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+using namespace std;
+
```
Comment (somya-kapoor) [inline Memory_management_algorithm/Best_fit.cpp:6]: Remove extra line

File: Memory_management_algorithm/Best_fit.cpp [hunk 35]
```diff
@@ -0,0 +1,103 @@
...
+
+void best_fit(int memory_block[], int process_size[], int m, int n) {
+     int allocated_list[n][3];
```
Comment (somya-kapoor) [inline Memory_management_algorithm/Best_fit.cpp:9]: You are using 8 space indentation. Use 4 space indentation

File: Memory_management_algorithm/Best_fit.cpp (hunk 32 above)
Comment (somya-kapoor) [reply Memory_management_algorithm/Best_fit.cpp:27]: Inconsistency in indentation

**Comment by somya-kapoor** [comment]
Due to no activity for more than 7 days, I am closing this PR

**Pull request closed without merging**